cpp/src/arrow/dataset/scanner.h

```diff
@@ -163,12 +164,20 @@ class ARROW_DS_EXPORT Scanner {
   /// in a concurrent fashion and outlive the iterator.
   Result<ScanTaskIterator> Scan();
 
+  /// \brief Apply a visitor to each RecordBatch as it is scanned. If multiple
+  /// threads are used, the visitor will be invoked from those threads and is
+  /// responsible for any synchronization.
+  Status Scan(std::function<Status(std::shared_ptr<RecordBatch>)> visitor);
```

Comment:
Why both `Scan(visitor)` and `ToBatches`?  Couldn't you just do `ToBatches().Visit(visitor)`?

Reply:
The difference is `ToBatches().Visit(visitor)` would invoke `visitor` exclusively on the thread which called `Visit()` whereas `Scan(visitor)` invokes `visitor` in the scan's thread pool. This method is speculative; I'm not sure we'd want to provide that but I included it as an example